### ISSUE TITLE
Add read-only chained references

### ISSUE THREAD

**Issue by nventuro** [balancer-labs/balancer-v2-monorepo#1507]
We've already thought of a number of scenarios in which they'd prove useful.

**Comment by nventuro** [comment]
We could have a second prefix that marks read-only refs, and e.g. have something like the following:

```
function _getChainedReferenceValue(uint256 ref) internal override returns (uint256 value) {
    (bytes32 slot, bool clear) = _getTempStorageSlot(ref);

    // Since we do manual calculation of storage slots, it is easier (and cheaper) to rely on internal assembly to
    // access it.

    // solhint-disable-next-line no-inline-assembly
    assembly {
        value := sload(slot)
    }

    if (clear) {
        sstore(slot, 0)   
    }
}
```

**Comment by jubeira** [comment]
Closed via #1664.
